lib/babe/build.go

```diff
@@ -189,11 +190,35 @@ func (b *BlockBuilder) buildBlockSeal(header *types.Header) (*types.SealDigest,
 func (b *BlockBuilder) buildBlockExtrinsics(slot Slot, rt runtime.Instance) []*transaction.ValidTransaction {
 	var included []*transaction.ValidTransaction
 
-	for !hasSlotEnded(slot) {
+	slotEnd := slot.start.Add(slot.duration * 2 / 3) // reserve last 1/3 of slot for block finalisation
+	timeout := time.Until(slotEnd)
+	slotTimer := time.NewTimer(timeout)
+
+	for {
+		select {
+		case <-b.testForceReturn:
+			if !slotTimer.Stop() {
+				<-slotTimer.C
+			}
+			return included
```

Comment:
You can remove that test channel + select case, but we need an integration test 😉

Reply:
Removed test channel and added integration test TestBuildBlock_addToTransactionState